lib/spack/spack/modules/common.py

```diff
@@ -269,13 +269,39 @@ def read_module_indices():
 module_type_to_indices = read_module_indices()
 
 
-def upstream_module(spec, module_type):
-    indices = module_type_to_indices[module_type]
+def upstream_module(spec, module_type, _module_indices=None):
+    indices = _module_indices or module_type_to_indices[module_type]
     for index in indices:
         if spec.dag_hash() in index:
             return index[spec.dag_hash()]
 
 
+def get_module(module_type, spec, get_full_path, module_indices=None):
+    if spec.package.installed_upstream:
+        module = spack.modules.common.upstream_module(
+            spec, module_type, _module_indices=module_indices)
+        if not module:
```

Comment:
Similar question here: what if there are multiple upstreams, or what if the module isn't available locally but *is* available in an upstream?  Which module should be returned?

It seems like there are gaps in the interface for cases where there are partial/different numbers of modules installed in chained Spack instances, but I suspect that's a case users will want to use `spack module <type> find` to debug.

Reply:
That is an issue: we should always attempt to retrieve the module associated with the Spack installation that installed the package. If a module is not available locally but we are using the local package, then we shouldn't use a module. Likewise, if we are instance X using upstreams Y and Z like X->Y->Z, then if we are using a package from instance Y, we should only use a module from instance Y.

As of now, the current, logic will only guarantee a correct result when a single upstream is used. To produce a correct result for additional upstream installations, the module index will have to be associated with the installation. That will take a bit of work but shouldn't take much longer than a day.